File: RB101_RB109_small_problems/easy_4/convert_a_string_to_a_number.rb
DIGITS = {
  "0" => 0,
  "1" => 1,
  "2" => 2,
  "3" => 3,
  "4" => 4,
  "5" => 5,
  "6" => 6,
  "7" => 7,
  "8" => 8,
  "9" => 9,
  "A" => 10,
  "B" => 11,
  "C" => 12,
  "D" => 13,
  "E" => 14,
  "F" => 15
}

def string_to_integer(string)
  string_array = string.split("")
  num = 0
  exponent = 0
  until string_array.empty?
    num += (DIGITS[string_array.pop] * (10 ** exponent))
    exponent += 1
  end
  num
end

p string_to_integer('4321') == 4321
p string_to_integer('570') == 570

# Further Exploration
def hexadecimal_to_integer(string)
  digits = string.upcase.chars.map { |char| DIGITS[char]}
  value = 0
  digits.each { |digit| value = 16 * value + digit }
  value
end

p hexadecimal_to_integer('4D9f') == 19871

File: JS210_fundamentals/basics/convert_a_string_to_a_number.js
const STRING_TO_DIGITS = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9];

function stringToInteger(str) {
  let digits = str.split('').reverse().map(d => STRING_TO_DIGITS[d]);

  let num = 0;

  for (let index = 0; index < digits.length; index += 1) {
    num += digits[index] * Math.pow(10, index);
  }

  return num;
}

console.log(stringToInteger('4321'));      // 4321
console.log(stringToInteger('570'));       // 570